// @progress/kendo-react-editor/dist/es/tools/table-wizard/tableProperties.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import * as React from 'react';
import { Button } from '@progress/kendo-react-buttons';
import { Checkbox, ColorPicker, NumericTextBox, TextBox } from '@progress/kendo-react-inputs';
import { AutoComplete, DropDownList } from '@progress/kendo-react-dropdowns';
import { Window, WindowActionsBar } from '@progress/kendo-react-dialogs';
import { TabStrip, TabStripTab } from '@progress/kendo-react-layout';
import { tableAlignMiddleCenterIcon, tableAlignMiddleLeftIcon, tableAlignMiddleRightIcon, tableAlignRemoveIcon, tablePropertiesIcon, tablePositionStartIcon, tablePositionCenterIcon, tablePositionEndIcon, positionTopIcon, positionBottomIcon } from '@progress/kendo-svg-icons';
import { borderStyles, onDownPreventDefault, popupSettings, units, parentNode } from './../utils';
import { itemRenderWithIcon, valueRenderWithIcon } from './tableCellProperties';
import { useLocalization } from '@progress/kendo-react-intl';
import { messages, keys } from './../../messages';
import { applyTableData, initialTableData, tableDefaultData, tablePositionStyles } from './tablePropsUtils';
/**
 * @hidden
 */
export var TableProperties = function (props) {
    var view = props.view, buttonProps = __rest(props, ["view"]);
    var loc = useLocalization();
    var _a = React.useState(false), showDialog = _a[0], setShowDialog = _a[1];
    var inTable = Boolean(view && parentNode(view.state.selection.$from, function (n) { return n.type.spec.tableRole === 'table'; }));
    var onClick = React.useCallback(function () {
        setShowDialog(true);
    }, []);
    var onClose = React.useCallback(function () {
        setShowDialog(false);
    }, []);
    var onSave = React.useCallback(function (data) {
        if (view && data) {
            var tr = applyTableData(view.state, data);
            if (tr) {
                view.dispatch(tr);
            }
        }
        setShowDialog(false);
    }, [view]);
    return (React.createElement(React.Fragment, null,
        React.createElement(Button, __assign({ type: "button", title: loc.toLanguageString(keys.tableProperties, messages[keys.tableProperties]), icon: 'table-properties', svgIcon: tablePropertiesIcon, onClick: onClick, disabled: !inTable }, onDownPreventDefault, buttonProps)),
        showDialog && React.createElement(TablePropertiesDialog, { onClose: onClose, onSave: onSave, view: view })));
};
var alignData = [
    { textKey: keys.tablePropertiesAlignLeft, text: messages[keys.tablePropertiesAlignLeft],
        value: 'left', icon: 'table-align-middle-left', svgIcon: tableAlignMiddleLeftIcon },
    { textKey: keys.tablePropertiesAlignCenter, text: messages[keys.tablePropertiesAlignCenter],
        value: 'center', icon: 'table-align-middle-center', svgIcon: tableAlignMiddleCenterIcon },
    { textKey: keys.tablePropertiesAlignRight, text: messages[keys.tablePropertiesAlignRight],
        value: 'right', icon: 'table-align-middle-right', svgIcon: tableAlignMiddleRightIcon },
    { textKey: keys.tablePropertiesNoAlignment, text: messages[keys.tablePropertiesNoAlignment],
        value: '', icon: 'align-remove', svgIcon: tableAlignRemoveIcon }
];
var tablePosition = [{
        textKey: keys.tablePropertiesLeft,
        text: messages[keys.tablePropertiesLeft],
        value: 'left',
        style: {},
        icon: 'table-position-left',
        svgIcon: tablePositionStartIcon
    }, {
        textKey: keys.tablePropertiesCenter,
        text: messages[keys.tablePropertiesCenter],
        value: 'center',
        style: {},
        icon: 'table-position-center',
        svgIcon: tablePositionCenterIcon
    }, {
        textKey: keys.tablePropertiesRight,
        text: messages[keys.tablePropertiesRight],
        value: 'right',
        style: {},
        icon: 'table-position-right',
        svgIcon: tablePositionEndIcon
    }].map(function (p) { return (__assign(__assign({}, p), { style: tablePositionStyles[p.value] })); });
var captionPositionData = [{
        textKey: keys.tablePropertiesTop,
        text: messages[keys.tablePropertiesTop],
        value: 'top',
        icon: 'position-top',
        svgIcon: positionTopIcon
    }, {
        textKey: keys.tablePropertiesBottom,
        text: messages[keys.tablePropertiesBottom],
        value: 'bottom',
        icon: 'position-bottom',
        svgIcon: positionBottomIcon
    }];
var associateHeadersData = [
    { textKey: keys.tablePropertiesNone, text: messages[keys.tablePropertiesNone], value: 'none' },
    { textKey: keys.tablePropertiesUsingScopeAttribute, text: messages[keys.tablePropertiesUsingScopeAttribute], value: 'scope' },
    { textKey: keys.tablePropertiesUsingIdAttributes, text: messages[keys.tablePropertiesUsingIdAttributes], value: 'id' }
];
var TablePropertiesDialog = function (props) {
    var defaults = React.useMemo(function () {
        var state = props.view && props.view.state;
        var data = state ? tableDefaultData(state) : initialTableData;
        return data;
    }, [props.view]);
    var loc = useLocalization();
    var _a = React.useState(0), selected = _a[0], setSelected = _a[1];
    var onSelect = React.useCallback(function (event) {
        setSelected(event.selected);
    }, []);
    // General tab refs
    var rows = React.useRef(null);
    var columns = React.useRef(null);
    var width = React.useRef(null);
    var height = React.useRef(null);
    var cellSpacing = React.useRef(null);
    var cellPadding = React.useRef(null);
    var borderWidth = React.useRef(null);
    var widthUnit = React.useRef(null);
    var heightUnit = React.useRef(null);
    var position = React.useRef(null);
    var alignment = React.useRef(null);
    var borderStyle = React.useRef(null);
    var backgroundColor = React.useRef(null);
    var borderColor = React.useRef(null);
    var collapseBorders = React.useRef(null);
    // Advanced tab refs
    var id = React.useRef(null);
    var className = React.useRef(null);
    var caption = React.useRef(null);
    var headerRows = React.useRef(null);
    var headerColumns = React.useRef(null);
    var captionAlignment = React.useRef(null);
    var captionPosition = React.useRef(null);
    var associateHeaders = React.useRef(null);
    var onSave = React.useCallback(function () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
        var data = {
            rows: ((_a = rows.current) === null || _a === void 0 ? void 0 : _a.value) || 0,
            columns: ((_b = columns.current) === null || _b === void 0 ? void 0 : _b.value) || 0,
            width: width.current ? width.current.value : null,
            height: height.current ? height.current.value : null,
            cellSpacing: cellSpacing.current ? cellSpacing.current.value : null,
            cellPadding: cellPadding.current ? cellPadding.current.value : null,
            borderWidth: borderWidth.current ? borderWidth.current.value : null,
            widthUnit: ((_c = widthUnit.current) === null || _c === void 0 ? void 0 : _c.value) || '',
            heightUnit: ((_d = heightUnit.current) === null || _d === void 0 ? void 0 : _d.value) || '',
            position: ((_f = (_e = position.current) === null || _e === void 0 ? void 0 : _e.value) === null || _f === void 0 ? void 0 : _f.value) || null,
            textAlign: ((_g = alignment.current) === null || _g === void 0 ? void 0 : _g.value) ? alignment.current.value.value : null,
            borderStyle: ((_h = borderStyle.current) === null || _h === void 0 ? void 0 : _h.value) ? borderStyle.current.value.value : null,
            backgroundColor: (_j = backgroundColor.current) === null || _j === void 0 ? void 0 : _j.value,
            borderColor: (_k = borderColor.current) === null || _k === void 0 ? void 0 : _k.value,
            collapseBorders: Boolean((_m = (_l = collapseBorders.current) === null || _l === void 0 ? void 0 : _l.element) === null || _m === void 0 ? void 0 : _m.checked),
            id: String(((_o = id.current) === null || _o === void 0 ? void 0 : _o.value) || ''),
            className: String(((_p = className.current) === null || _p === void 0 ? void 0 : _p.value) || ''),
            caption: String(((_q = caption.current) === null || _q === void 0 ? void 0 : _q.value) || ''),
            headerRows: ((_r = headerRows.current) === null || _r === void 0 ? void 0 : _r.value) || 0,
            headerColumns: ((_s = headerColumns.current) === null || _s === void 0 ? void 0 : _s.value) || 0,
            captionAlignment: ((_t = captionAlignment.current) === null || _t === void 0 ? void 0 : _t.value) ? captionAlignment.current.value.value : null,
            captionPosition: ((_u = captionPosition.current) === null || _u === void 0 ? void 0 : _u.value) ? captionPosition.current.value.value : null,
            associateHeaders: ((_v = associateHeaders.current) === null || _v === void 0 ? void 0 : _v.value) ? associateHeaders.current.value.value : 'none'
        };
        props.onSave.call(undefined, data);
    }, [props.onSave]);
    var general = (React.createElement("div", { className: "k-form k-form-md" },
        React.createElement("fieldset", { className: "k-form-fieldset" },
            React.createElement("div", { className: "k-form-layout k-d-grid k-grid-cols-4 k-gap-x-4" },
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesRows, messages[keys.tablePropertiesRows])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(NumericTextBox, { ref: rows, defaultValue: defaults.rows, min: 1 }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesColumns, messages[keys.tablePropertiesColumns])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(NumericTextBox, { ref: columns, defaultValue: defaults.columns, min: 1 }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesWidth, messages[keys.tablePropertiesWidth])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(NumericTextBox, { ref: width, defaultValue: defaults.width, min: 0, placeholder: loc.toLanguageString(keys.sizeAuto, messages[keys.sizeAuto]) }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, "\u00A0"),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(AutoComplete, { ref: widthUnit, defaultValue: defaults.widthUnit, data: units, popupSettings: popupSettings }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesHeight, messages[keys.tablePropertiesHeight])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(NumericTextBox, { ref: height, defaultValue: defaults.height, min: 0, placeholder: loc.toLanguageString(keys.sizeAuto, messages[keys.sizeAuto]) }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, "\u00A0"),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(AutoComplete, { ref: heightUnit, defaultValue: defaults.heightUnit, data: units, popupSettings: popupSettings }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesPosition, messages[keys.tablePropertiesPosition])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(DropDownList, { ref: position, defaultValue: tablePosition.find(function (p) { return p.value === defaults.position; }), popupSettings: popupSettings, data: tablePosition.map(function (i) { return (__assign(__assign({}, i), { text: loc.toLanguageString(i.textKey, i.text) })); }), textField: 'text', dataItemKey: 'value', valueRender: valueRenderWithIcon, itemRender: itemRenderWithIcon }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesAlignment, messages[keys.tablePropertiesAlignment])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(DropDownList, { ref: alignment, defaultValue: alignData.find(function (a) { return a.value === defaults.textAlign; }), data: alignData.map(function (i) { return (__assign(__assign({}, i), { text: loc.toLanguageString(i.textKey, i.text) })); }), textField: 'text', dataItemKey: 'value', popupSettings: popupSettings, valueRender: valueRenderWithIcon, itemRender: itemRenderWithIcon }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesCellSpacing, messages[keys.tablePropertiesCellSpacing])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(NumericTextBox, { ref: cellSpacing, defaultValue: defaults.cellSpacing, placeholder: loc.toLanguageString(keys.sizeAuto, messages[keys.sizeAuto]), min: 0 }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesCellPadding, messages[keys.tablePropertiesCellPadding])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(NumericTextBox, { ref: cellPadding, defaultValue: defaults.cellPadding, placeholder: loc.toLanguageString(keys.sizeAuto, messages[keys.sizeAuto]), min: 0 }))),
                React.createElement("div", { className: "k-form-field k-col-span-full" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesTableBackground, messages[keys.tablePropertiesTableBackground])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(ColorPicker, { ref: backgroundColor, defaultValue: defaults.backgroundColor }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesBorderWidth, messages[keys.tablePropertiesBorderWidth])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(NumericTextBox, { ref: borderWidth, defaultValue: defaults.borderWidth, min: 0 }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesBorderColor, messages[keys.tablePropertiesBorderColor])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(ColorPicker, { ref: borderColor, defaultValue: defaults.borderColor }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesBorderStyle, messages[keys.tablePropertiesBorderStyle])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(DropDownList, { ref: borderStyle, defaultValue: borderStyles.find(function (s) { return s.value === defaults.borderStyle; }), popupSettings: popupSettings, data: borderStyles.map(function (i) { return (__assign(__assign({}, i), { text: loc.toLanguageString(i.textKey, i.text) })); }), textField: 'text', dataItemKey: 'value' }))),
                React.createElement("div", { className: "k-form-field k-col-span-full" },
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(Checkbox, { ref: collapseBorders, className: "k-checkbox-wrap", label: loc.toLanguageString(keys.tablePropertiesCollapseBorders, messages[keys.tablePropertiesCollapseBorders]), defaultChecked: defaults.collapseBorders })))))));
    var advanced = (React.createElement("div", { className: "k-form k-form-md" },
        React.createElement("fieldset", { className: "k-form-fieldset" },
            React.createElement("div", { className: "k-form-layout k-d-grid" },
                React.createElement("div", { className: "k-form-field" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesId, messages[keys.tablePropertiesId])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(TextBox, { ref: id, defaultValue: defaults.id }))),
                React.createElement("div", { className: "k-form-field" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesCssClass, messages[keys.tablePropertiesCssClass])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(TextBox, { ref: className, defaultValue: defaults.className }))),
                React.createElement("fieldset", { className: "k-form-fieldset" },
                    React.createElement("legend", { className: "k-form-legend" }, loc.toLanguageString(keys.tablePropertiesAccessibility, messages[keys.tablePropertiesAccessibility])),
                    React.createElement("div", { className: "k-d-grid k-grid-cols-4 k-gap-x-4" },
                        React.createElement("div", { className: "k-form-field k-col-span-full" },
                            React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesCaption, messages[keys.tablePropertiesCaption])),
                            React.createElement("div", { className: "k-form-field-wrap" },
                                React.createElement(TextBox, { ref: caption, defaultValue: defaults.caption }))),
                        React.createElement("div", { className: "k-form-field k-col-span-2" },
                            React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesCaptionPosition, messages[keys.tablePropertiesCaptionPosition])),
                            React.createElement("div", { className: "k-form-field-wrap" },
                                React.createElement(DropDownList, { ref: captionPosition, defaultValue: captionPositionData.find(function (i) { return i.value === defaults.captionPosition; }), data: captionPositionData.map(function (i) { return (__assign(__assign({}, i), { text: loc.toLanguageString(i.textKey, i.text) })); }), popupSettings: popupSettings, textField: 'text', dataItemKey: 'value', valueRender: valueRenderWithIcon, itemRender: itemRenderWithIcon }))),
                        React.createElement("div", { className: "k-form-field k-col-span-2" },
                            React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesCaptionAlignment, messages[keys.tablePropertiesCaptionAlignment])),
                            React.createElement("div", { className: "k-form-field-wrap" },
                                React.createElement(DropDownList, { ref: captionAlignment, defaultValue: alignData.find(function (i) { return i.value === defaults.captionAlignment; }), data: alignData.map(function (i) { return (__assign(__assign({}, i), { text: loc.toLanguageString(i.textKey, i.text) })); }), popupSettings: popupSettings, textField: 'text', dataItemKey: 'value', valueRender: valueRenderWithIcon, itemRender: itemRenderWithIcon }))),
                        React.createElement("div", { className: "k-form-field k-col-span-1" },
                            React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesHeaderRows, messages[keys.tablePropertiesHeaderRows])),
                            React.createElement("div", { className: "k-form-field-wrap" },
                                React.createElement(NumericTextBox, { ref: headerRows, defaultValue: defaults.headerRows, min: 0 }))),
                        React.createElement("div", { className: "k-form-field k-col-span-1" },
                            React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesHeaderCols, messages[keys.tablePropertiesHeaderCols])),
                            React.createElement("div", { className: "k-form-field-wrap" },
                                React.createElement(NumericTextBox, { ref: headerColumns, defaultValue: defaults.headerColumns, min: 0 }))),
                        React.createElement("div", { className: "k-form-field k-col-span-2" },
                            React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(keys.tablePropertiesAssociateHeaders, messages[keys.tablePropertiesAssociateHeaders])),
                            React.createElement("div", { className: "k-form-field-wrap" },
                                React.createElement(DropDownList, { ref: associateHeaders, defaultValue: associateHeadersData.find(function (i) { return i.value === defaults.associateHeaders; }), data: associateHeadersData.map(function (i) { return (__assign(__assign({}, i), { text: loc.toLanguageString(i.textKey, i.text) })); }), popupSettings: popupSettings, textField: 'text', dataItemKey: 'value' })))))))));
    var content = (React.createElement(TabStrip, { selected: selected, onSelect: onSelect, animation: false, keepTabsMounted: true },
        React.createElement(TabStripTab, { title: loc.toLanguageString(keys.tablePropertiesGeneral, messages[keys.tablePropertiesGeneral]) }, general),
        React.createElement(TabStripTab, { title: loc.toLanguageString(keys.tablePropertiesAdvanced, messages[keys.tablePropertiesAdvanced]) }, advanced)));
    var actionButtons = [(React.createElement(Button, { themeColor: 'primary', key: "save", onClick: onSave }, loc.toLanguageString(keys.tablePropertiesSave, messages[keys.tablePropertiesSave]))), (React.createElement(Button, { onClick: props.onClose, key: "cancel" }, loc.toLanguageString(keys.tablePropertiesCancel, messages[keys.tablePropertiesCancel])))];
    return (React.createElement(Window, { className: 'k-editor-window k-editor-table-wizard-window', title: loc.toLanguageString(keys.tableProperties, messages[keys.tableProperties]), onClose: props.onClose, key: "dialog", style: { userSelect: 'none' }, width: 480, height: 630, modal: true, minimizeButton: function () { return null; }, maximizeButton: function () { return null; }, resizable: false },
        content,
        React.createElement.apply(React, __spreadArray([WindowActionsBar, { layout: 'start' }], actionButtons, false))));
};
