// @progress/kendo-react-editor/dist/npm/tools/table-wizard/tableCellProperties.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.itemRenderWithIcon = exports.valueRenderWithIcon = exports.TableCellProperties = void 0;
var React = require("react");
var kendo_react_buttons_1 = require("@progress/kendo-react-buttons");
var kendo_react_inputs_1 = require("@progress/kendo-react-inputs");
var kendo_react_dropdowns_1 = require("@progress/kendo-react-dropdowns");
var kendo_react_dialogs_1 = require("@progress/kendo-react-dialogs");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
var utils_1 = require("../utils");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var kendo_react_intl_1 = require("@progress/kendo-react-intl");
var messages_1 = require("./../../messages");
var cellPropsUtils_1 = require("./cellPropsUtils");
/**
 * @hidden
 */
var TableCellProperties = function (props) {
    var view = props.view, buttonProps = __rest(props, ["view"]);
    var _a = React.useState(false), showDialog = _a[0], setShowDialog = _a[1];
    var loc = (0, kendo_react_intl_1.useLocalization)();
    var inTable = Boolean(view && (0, utils_1.parentNode)(view.state.selection.$from, function (n) { return n.type.spec.tableRole === 'table'; }));
    var onClick = React.useCallback(function () {
        setShowDialog(true);
    }, []);
    var onSave = React.useCallback(function (data) {
        if (view && data) {
            var tr = (0, cellPropsUtils_1.applyCellsData)(view.state, data);
            if (tr.docChanged) {
                view.dispatch(tr);
            }
        }
        setShowDialog(false);
    }, [view]);
    var onCancel = React.useCallback(function () {
        setShowDialog(false);
    }, []);
    return (React.createElement(React.Fragment, null,
        React.createElement(kendo_react_buttons_1.Button, __assign({ type: "button", title: loc.toLanguageString(messages_1.keys.cellProperties, messages_1.messages[messages_1.keys.cellProperties]), icon: 'table-cell-properties', svgIcon: kendo_svg_icons_1.tableCellPropertiesIcon, onClick: onClick, disabled: !inTable }, utils_1.onDownPreventDefault, buttonProps)),
        showDialog && view && React.createElement(TableCellPropertiesDialog, { view: view, onCancel: onCancel, onSave: onSave })));
};
exports.TableCellProperties = TableCellProperties;
var alignData = [
    { textKey: messages_1.keys.cellPropertiesLeftTop, text: messages_1.messages[messages_1.keys.cellPropertiesLeftTop],
        icon: 'table-align-top-left', svgIcon: kendo_svg_icons_1.tableAlignTopLeftIcon, value: { x: 'left', y: 'top' } },
    { textKey: messages_1.keys.cellPropertiesCenterTop, text: messages_1.messages[messages_1.keys.cellPropertiesCenterTop],
        icon: 'table-align-top-center', svgIcon: kendo_svg_icons_1.tableAlignTopCenterIcon, value: { x: 'center', y: 'top' } },
    { textKey: messages_1.keys.cellPropertiesRightTop, text: messages_1.messages[messages_1.keys.cellPropertiesRightTop],
        icon: 'table-align-top-right', svgIcon: kendo_svg_icons_1.tableAlignTopRightIcon, value: { x: 'right', y: 'top' } },
    { textKey: messages_1.keys.cellPropertiesLeftMiddle, text: messages_1.messages[messages_1.keys.cellPropertiesLeftMiddle],
        icon: 'table-align-middle-left', svgIcon: kendo_svg_icons_1.tableAlignMiddleLeftIcon, value: { x: 'left', y: 'middle' } },
    { textKey: messages_1.keys.cellPropertiesCenterMiddle, text: messages_1.messages[messages_1.keys.cellPropertiesCenterMiddle],
        icon: 'table-align-middle-center', svgIcon: kendo_svg_icons_1.tableAlignMiddleCenterIcon, value: { x: 'center', y: 'middle' } },
    { textKey: messages_1.keys.cellPropertiesRightMiddle, text: messages_1.messages[messages_1.keys.cellPropertiesRightMiddle],
        icon: 'table-align-middle-right', svgIcon: kendo_svg_icons_1.tableAlignMiddleRightIcon, value: { x: 'right', y: 'middle' } },
    { textKey: messages_1.keys.cellPropertiesLeftBottom, text: messages_1.messages[messages_1.keys.cellPropertiesLeftBottom],
        icon: 'table-align-bottom-left', svgIcon: kendo_svg_icons_1.tableAlignBottomLeftIcon, value: { x: 'left', y: 'bottom' } },
    { textKey: messages_1.keys.cellPropertiesCenterBottom, text: messages_1.messages[messages_1.keys.cellPropertiesCenterBottom],
        icon: 'table-align-bottom-center', svgIcon: kendo_svg_icons_1.tableAlignBottomCenterIcon, value: { x: 'center', y: 'bottom' } },
    { textKey: messages_1.keys.cellPropertiesRightBottom, text: messages_1.messages[messages_1.keys.cellPropertiesRightBottom],
        icon: 'table-align-bottom-right', svgIcon: kendo_svg_icons_1.tableAlignBottomRightIcon, value: { x: 'right', y: 'bottom' } },
    { textKey: messages_1.keys.cellPropertiesNoAlignment, text: messages_1.messages[messages_1.keys.cellPropertiesNoAlignment],
        icon: 'align-remove', svgIcon: kendo_svg_icons_1.tableAlignRemoveIcon, value: { x: '', y: '' } }
];
var textControlData = [
    { textKey: messages_1.keys.cellPropertiesWrap, text: messages_1.messages[messages_1.keys.cellPropertiesWrap],
        value: '', icon: 'text-wrap', svgIcon: kendo_svg_icons_1.textWrapIcon },
    { textKey: messages_1.keys.cellPropertiesFitToCell, text: messages_1.messages[messages_1.keys.cellPropertiesFitToCell],
        value: 'nowrap', icon: 'parameter-string', svgIcon: kendo_svg_icons_1.parameterStringIcon }
];
/**
 * @hidden
 */
var valueRenderWithIcon = function (element, value) {
    if (!value) {
        return element;
    }
    return (React.createElement(React.Fragment, null,
        React.createElement(kendo_react_inputs_1.InputPrefix, null,
            React.createElement(kendo_react_common_1.IconWrap, { icon: value.svgIcon, name: value.icon })),
        element));
};
exports.valueRenderWithIcon = valueRenderWithIcon;
/**
 * @hidden
 */
var itemRenderWithIcon = function (li, itemProps) {
    var itemChildren = (React.createElement(React.Fragment, null,
        React.createElement(kendo_react_inputs_1.InputPrefix, null,
            React.createElement(kendo_react_common_1.IconWrap, { icon: itemProps.dataItem.svgIcon, name: itemProps.dataItem.icon })),
        li.props.children));
    return React.cloneElement(li, li.props, itemChildren);
};
exports.itemRenderWithIcon = itemRenderWithIcon;
var TableCellPropertiesDialog = function (props) {
    var defaults = React.useMemo(function () {
        return (0, cellPropsUtils_1.cellsDefaultData)(props.view.state);
    }, [props.view]);
    var loc = (0, kendo_react_intl_1.useLocalization)();
    var applyToAll = React.useRef(null);
    var applyToAllDef = defaults.applyToAll;
    var cellWidth = React.useRef(null);
    var cellWidthDef = defaults.cellWidth;
    var widthUnit = React.useRef(null);
    if (defaults.widthUnit && !utils_1.units.includes(defaults.widthUnit)) {
        utils_1.units.push(defaults.widthUnit);
    }
    var widthUnitDef = defaults.widthUnit || '';
    var cellHeight = React.useRef(null);
    var cellHeightDef = defaults.cellHeight;
    var heightUnit = React.useRef(null);
    if (defaults.heightUnit && !utils_1.units.includes(defaults.heightUnit)) {
        utils_1.units.push(defaults.heightUnit);
    }
    var heightUnitDef = defaults.heightUnit || '';
    var alignment = React.useRef(null);
    var alignmentDef = alignData.find(function (al) { return al.value.x !== undefined && defaults.alignment &&
        al.value.x === defaults.alignment.x && al.value.y === defaults.alignment.y; }) || null;
    var textControl = React.useRef(null);
    var textControlDef = textControlData.find(function (d) { return d.value === defaults.textControl; });
    var backgroundColor = React.useRef(null);
    var backgroundColorDef = defaults.backgroundColor;
    var cellPadding = React.useRef(null);
    var cellPaddingDef = defaults.cellPadding;
    var borderWidth = React.useRef(null);
    var borderWidthDef = defaults.borderWidth;
    var borderColor = React.useRef(null);
    var borderColorDef = defaults.borderColor;
    var borderStyle = React.useRef(null);
    var borderStyleDef = utils_1.borderStyles.find(function (s) { return s.value === defaults.borderStyle; });
    var id = React.useRef(null);
    var idDef = defaults.id;
    var className = React.useRef(null);
    var classNameDef = defaults.className;
    var onSave = React.useCallback(function () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
        var data = {
            applyToAll: Boolean((_b = (_a = applyToAll.current) === null || _a === void 0 ? void 0 : _a.element) === null || _b === void 0 ? void 0 : _b.checked),
            cellWidth: ((_c = cellWidth.current) === null || _c === void 0 ? void 0 : _c.value) || null,
            widthUnit: ((_d = widthUnit.current) === null || _d === void 0 ? void 0 : _d.value) || '',
            cellHeight: ((_e = cellHeight.current) === null || _e === void 0 ? void 0 : _e.value) || null,
            heightUnit: ((_f = heightUnit.current) === null || _f === void 0 ? void 0 : _f.value) || '',
            alignment: (_h = (_g = alignment.current) === null || _g === void 0 ? void 0 : _g.value) === null || _h === void 0 ? void 0 : _h.value,
            textControl: (_k = (_j = textControl.current) === null || _j === void 0 ? void 0 : _j.value) === null || _k === void 0 ? void 0 : _k.value,
            backgroundColor: (_l = backgroundColor.current) === null || _l === void 0 ? void 0 : _l.value,
            cellPadding: typeof ((_m = cellPadding.current) === null || _m === void 0 ? void 0 : _m.value) === 'number' ? (_o = cellPadding.current) === null || _o === void 0 ? void 0 : _o.value : null,
            borderWidth: typeof ((_p = borderWidth.current) === null || _p === void 0 ? void 0 : _p.value) === 'number' ? (_q = borderWidth.current) === null || _q === void 0 ? void 0 : _q.value : null,
            borderColor: (_r = borderColor.current) === null || _r === void 0 ? void 0 : _r.value,
            borderStyle: ((_t = (_s = borderStyle.current) === null || _s === void 0 ? void 0 : _s.value) === null || _t === void 0 ? void 0 : _t.value) || null,
            id: String(((_u = id.current) === null || _u === void 0 ? void 0 : _u.value) || ''),
            className: String(((_v = className.current) === null || _v === void 0 ? void 0 : _v.value) || '')
        };
        props.onSave.call(undefined, data);
    }, [props.onSave]);
    var content = (React.createElement("div", { className: "k-form k-form-md" },
        React.createElement("fieldset", { className: "k-form-fieldset" },
            React.createElement("div", { className: "k-form-layout k-d-grid k-grid-cols-4 k-gap-x-4" },
                React.createElement("div", { className: "k-form-field k-col-span-full" },
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement("span", { className: "k-checkbox-wrap" },
                            React.createElement(kendo_react_inputs_1.Checkbox, { ref: applyToAll, defaultChecked: applyToAllDef })),
                        React.createElement("label", { className: "k-checkbox-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesApplyToAllCells, messages_1.messages[messages_1.keys.cellPropertiesApplyToAllCells])))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesWidth, messages_1.messages[messages_1.keys.cellPropertiesWidth])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_inputs_1.NumericTextBox, { ref: cellWidth, defaultValue: cellWidthDef, min: 0, placeholder: loc.toLanguageString(messages_1.keys.sizeAuto, messages_1.messages[messages_1.keys.sizeAuto]) }),
                        React.createElement("div", { className: "k-form-hint" }, loc.toLanguageString(messages_1.keys.cellPropertiesApplyToColumn, messages_1.messages[messages_1.keys.cellPropertiesApplyToColumn])))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, "\u00A0"),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_dropdowns_1.AutoComplete, { ref: widthUnit, defaultValue: widthUnitDef, data: utils_1.units, popupSettings: utils_1.popupSettings }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesHeight, messages_1.messages[messages_1.keys.cellPropertiesHeight])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_inputs_1.NumericTextBox, { ref: cellHeight, defaultValue: cellHeightDef, min: 0, placeholder: loc.toLanguageString(messages_1.keys.sizeAuto, messages_1.messages[messages_1.keys.sizeAuto]) }),
                        React.createElement("div", { className: "k-form-hint" }, loc.toLanguageString(messages_1.keys.cellPropertiesApplyToRow, messages_1.messages[messages_1.keys.cellPropertiesApplyToRow])))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, "\u00A0"),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_dropdowns_1.AutoComplete, { ref: heightUnit, defaultValue: heightUnitDef, data: utils_1.units, popupSettings: utils_1.popupSettings }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesAlignment, messages_1.messages[messages_1.keys.cellPropertiesAlignment])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_dropdowns_1.DropDownList, { ref: alignment, defaultValue: alignmentDef, data: alignData.map(function (i) { return (__assign(__assign({}, i), { text: loc.toLanguageString(i.textKey, i.text) })); }), textField: 'text', dataItemKey: 'text', popupSettings: utils_1.popupSettings, valueRender: exports.valueRenderWithIcon, itemRender: exports.itemRenderWithIcon }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesTextControl, messages_1.messages[messages_1.keys.cellPropertiesTextControl])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_dropdowns_1.DropDownList, { ref: textControl, defaultValue: textControlDef, data: textControlData.map(function (i) { return (__assign(__assign({}, i), { text: loc.toLanguageString(i.textKey, i.text) })); }), textField: 'text', dataItemKey: 'value', popupSettings: utils_1.popupSettings, valueRender: exports.valueRenderWithIcon, itemRender: exports.itemRenderWithIcon }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesCellPadding, messages_1.messages[messages_1.keys.cellPropertiesCellPadding])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_inputs_1.NumericTextBox, { ref: cellPadding, defaultValue: cellPaddingDef, min: 0 }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesBackground, messages_1.messages[messages_1.keys.cellPropertiesBackground])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_inputs_1.ColorPicker, { ref: backgroundColor, defaultValue: backgroundColorDef }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesBorderWidth, messages_1.messages[messages_1.keys.cellPropertiesBorderWidth])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_inputs_1.NumericTextBox, { ref: borderWidth, defaultValue: borderWidthDef, min: 0 }))),
                React.createElement("div", { className: "k-form-field k-col-span-1" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesBorderColor, messages_1.messages[messages_1.keys.cellPropertiesBorderColor])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_inputs_1.ColorPicker, { ref: borderColor, defaultValue: borderColorDef }))),
                React.createElement("div", { className: "k-form-field k-col-span-2" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesBorderStyle, messages_1.messages[messages_1.keys.cellPropertiesBorderStyle])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_dropdowns_1.DropDownList, { ref: borderStyle, defaultValue: borderStyleDef, popupSettings: utils_1.popupSettings, data: utils_1.borderStyles, textField: 'text', dataItemKey: 'value' }))),
                React.createElement("div", { className: "k-form-field k-col-span-full" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesId, messages_1.messages[messages_1.keys.cellPropertiesId])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_inputs_1.TextBox, { ref: id, defaultValue: idDef }))),
                React.createElement("div", { className: "k-form-field k-col-span-full" },
                    React.createElement("label", { className: "k-label k-form-label" }, loc.toLanguageString(messages_1.keys.cellPropertiesCssClass, messages_1.messages[messages_1.keys.cellPropertiesCssClass])),
                    React.createElement("div", { className: "k-form-field-wrap" },
                        React.createElement(kendo_react_inputs_1.TextBox, { ref: className, defaultValue: classNameDef })))))));
    var actionButtons = [(React.createElement(kendo_react_buttons_1.Button, { themeColor: 'primary', key: "save", onClick: onSave }, loc.toLanguageString(messages_1.keys.cellPropertiesSave, messages_1.messages[messages_1.keys.cellPropertiesSave]))), (React.createElement(kendo_react_buttons_1.Button, { onClick: props.onCancel, key: "cancel" }, loc.toLanguageString(messages_1.keys.cellPropertiesCancel, messages_1.messages[messages_1.keys.cellPropertiesCancel])))];
    return (React.createElement(kendo_react_dialogs_1.Window, { className: 'k-editor-window k-editor-table-wizard-window', title: loc.toLanguageString(messages_1.keys.cellProperties, messages_1.messages[messages_1.keys.cellProperties]), onClose: props.onCancel, key: "dialog", style: { userSelect: 'none' }, width: 408, height: 587, modal: true, minimizeButton: function () { return null; }, maximizeButton: function () { return null; }, resizable: false },
        content,
        React.createElement.apply(React, __spreadArray([kendo_react_dialogs_1.WindowActionsBar, { layout: 'start' }], actionButtons, false))));
};
