// @progress/kendo-react-dropdowns/dist/npm/messages/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.adaptiveModeFooterCancel = exports.adaptiveModeFooterApply = exports.dropDownListArrowBtnAriaLabel = exports.comboArrowBtnAriaLabelCollapse = exports.comboArrowBtnAriaLabelExpand = exports.clear = exports.nodata = void 0;
/**
 * @hidden
 */
exports.nodata = 'dropdowns.nodata';
/**
 * @hidden
 */
exports.clear = 'dropdowns.clear';
/**
 * @hidden
 */
exports.comboArrowBtnAriaLabelExpand = 'dropdowns.comboArrowBtnAriaLabelExpand';
/**
 * @hidden
 */
exports.comboArrowBtnAriaLabelCollapse = 'dropdowns.comboArrowBtnAriaLabelCollapse';
/**
 * @hidden
 */
exports.dropDownListArrowBtnAriaLabel = 'dropdowns.dropDownListArrowBtnAriaLabel';
/**
 * @hidden
 */
exports.adaptiveModeFooterApply = 'dropdowns.apply';
/**
 * @hidden
 */
exports.adaptiveModeFooterCancel = 'dropdowns.cancel';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.clear] = 'clear',
    _a[exports.nodata] = 'NO DATA FOUND.',
    _a[exports.comboArrowBtnAriaLabelExpand] = 'expand combobox',
    _a[exports.comboArrowBtnAriaLabelCollapse] = 'collapse combobox',
    _a[exports.dropDownListArrowBtnAriaLabel] = 'select',
    _a[exports.adaptiveModeFooterApply] = 'Apply',
    _a[exports.adaptiveModeFooterCancel] = 'Cancel',
    _a);
